<a href="https://colab.research.google.com/github/aikssa/python_do_zero_ao_ds/blob/main/aula_01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Análise de preços de casas - Aula 01

# Importação das bibliotecas e dos dados

In [1]:
# importando as bibliotecas necessárias
import pandas as pd

In [2]:
# carregando o dataset
df = pd.read_csv('/content/drive/MyDrive/repos/python_zero_ds/datasets/kc_house_data.csv')

# verificando as 5 primeiras linhas do dataframe
df.head()

id             date     price  ...     long  sqft_living15  sqft_lot15
0  7129300520  20141013T000000  221900.0  ... -122.257           1340        5650
1  6414100192  20141209T000000  538000.0  ... -122.319           1690        7639
2  5631500400  20150225T000000  180000.0  ... -122.233           2720        8062
3  2487200875  20141209T000000  604000.0  ... -122.393           1360        5000
4  1954400510  20150218T000000  510000.0  ... -122.045           1800        7503

[5 rows x 21 columns]

# Limpeza dos dados

Inicialmente vamos verificar a existência de dados duplicado, ou seja, aqueles com o mesmo `id`.

In [3]:
# verificando a quantidade de dados duplicados
df['id'].duplicated().value_counts()

False    21436
True       177
Name: id, dtype: int64

É possível verificar que existem 177 campos na coluna `id` que estão duplicados.

Podemos supor que isso acontece em razão de uma atualização do preço, sem a exclusão do lançamento anterior. Isso pode ser comprovado pelos dois exemplos abaixo, onde apenas as colunas `date` e `price` estão diferentes.

In [4]:
# verificando os campos duplicados para selecionarmos alguns exemplos
df[df['id'].duplicated() == True].head()

id             date      price  ...     long  sqft_living15  sqft_lot15
94   6021501535  20141223T000000   700000.0  ... -122.386           1570        4500
314  4139480200  20141209T000000  1400000.0  ... -122.102           3860       11244
325  7520000520  20150311T000000   240500.0  ... -122.352           1820        7460
346  3969300030  20141229T000000   239900.0  ... -122.240           1020        7138
372  2231500030  20150324T000000   530000.0  ... -122.341           1810        6929

[5 rows x 21 columns]

In [5]:
# mostrando o primeiro exemplo
df[df['id'] == 6021501535]

id             date     price  ...     long  sqft_living15  sqft_lot15
93  6021501535  20140725T000000  430000.0  ... -122.386           1570        4500
94  6021501535  20141223T000000  700000.0  ... -122.386           1570        4500

[2 rows x 21 columns]

In [6]:
# mostrando o segundo exemplo
df[df['id'] == 3969300030]

id             date     price  ...    long  sqft_living15  sqft_lot15
345  3969300030  20140723T000000  165000.0  ... -122.24           1020        7138
346  3969300030  20141229T000000  239900.0  ... -122.24           1020        7138

[2 rows x 21 columns]

Deste modo, iremos remover as linhas duplicadas, optando por deixar apenas a mais recente.

In [7]:
# removendo as linhas duplicadas
df.drop_duplicates(subset='id', keep='last', inplace=True)

# verificando o resultado
df.duplicated().value_counts()

False    21436
dtype: int64

Note que agora não existem mais valores duplicados.

Abaixo vamos confirmar isso a partir dos exemplos selecionados anteriormente.

In [8]:
# verificando o primeiro exemplo
df[df['id'] == 6021501535]

id             date     price  ...     long  sqft_living15  sqft_lot15
94  6021501535  20141223T000000  700000.0  ... -122.386           1570        4500

[1 rows x 21 columns]

In [9]:
# verificando o segundo exemplo
df[df['id'] == 3969300030]

id             date     price  ...    long  sqft_living15  sqft_lot15
346  3969300030  20141229T000000  239900.0  ... -122.24           1020        7138

[1 rows x 21 columns]

Após essa limpeza inicial dos dados, vamos responder alguns questionamentos.

## Q1: Quantas casas estão disponíveis para compra?

Com a remoção dos valores duplicados, os `id` agora são únicos e, portanto, descobrindo a quantidade de linhas, é possível determinar a quantidade de casas que estão disponíveis para compra.

In [10]:
# verificando a quantidade de linhas
df.shape[0]

21436

**Resposta**: Existem 21.436 imóveis disponíveis para compra.

## Q2: Quantos atributos as casas possuem?

In [11]:
# verificando a quantidade de colunas
df.shape[1]

21

**Respostas**: Os imóveis possuem 21 atributos.

## Q3: Quais são os atributos das casas?

In [12]:
# verificando o nome das colunas
df.columns

Index(['id', 'date', 'price', 'bedrooms', 'bathrooms', 'sqft_living',
       'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade',
       'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode',
       'lat', 'long', 'sqft_living15', 'sqft_lot15'],
      dtype='object')

**Resposta**: Os 21 atributos dos imóveis são: id, data de venda, preço de venda, número de quartos e de banheiros, metragem em pés da área interna e da área externa, a quantidade de andares, se tem ou não vista para o mar, a qualidade da vista do imóvel, a condição de conservação do imóvel, a qualidade da construção e do design, a metragem em pés do espaço interno acima e abaixo do nível do solo, ano de construção e da última reforma, CEP, latitude, longitude e metragem em pés das áreas internas e externas dos 15 vizinhos mais próximos.

## Q4: Qual a casa mais cara (casa com o maior valor de venda)?

In [13]:
# ordenando o preço de maneira decrescente e mostrando os 2 primeiros registros
df.sort_values('price', ascending=False).head(2)

id             date  ...  sqft_living15  sqft_lot15
7252  6762700020  20141013T000000  ...           3940        8800
3914  9808700762  20140611T000000  ...           3930       25449

[2 rows x 21 columns]

**Resposta**: A casa mais cara é a 6762700020, custando $ 7.700.000,00.

## Q5: Qual a casa com o maior número de quartos?

In [14]:
# ordenando o número de quartos de maneira decrescente e mostrando os 2 primeiros registros
df.sort_values('bedrooms', ascending=False).head(2)

id             date  ...  sqft_living15  sqft_lot15
15870  2402100895  20140625T000000  ...           1330        4700
8757   1773100755  20140821T000000  ...           1420        4960

[2 rows x 21 columns]

**Resposta**: A casa com o maior número de quartos é a 2402100895, com 33 quartos.

Aqui é possível verificar que, embora o número de quartos seja alto, o preço, principalmente se comparado à casa mais cara, é bem baixo. Isso pode significar que esse dado é um outlier, ou seja, difere muito do restante da base de dados.

## Q6: Qual a soma total de quartos do conjunto de dados?

In [15]:
# somando a quantidade total de quartos
df['bedrooms'].sum()

72273

**Resposta**: A soma total de quartos do conjunto de dados é 72.273.

## Q7: Quantas casas possuem 2 banheiros?

In [16]:
# verificando quantas casas possuem 2 banheiros
df.loc[df['bathrooms'] == 2].shape[0]

1913

**Resposta**: 1.913 imóveis possuem dois banheiros.

## Q8: Qual o preço médio de todas as casas no conjunto de dados?

In [17]:
# verificando a média dos preços
df['price'].mean()

541649.962726255

**Resposta**: O preço médio de todos os imóveis do conjunto de dados é $ 541.649,96.

## Q9: Qual o preço médio de casas com 2 banheiros?

In [18]:
# verificando o preço médio das casas com 2 banheiros
df[['price', 'bathrooms']].loc[df['bathrooms'] == 2].mean()

price        459307.013591
bathrooms         2.000000
dtype: float64

**Resposta**: O preço médio dos imóveis com 2 banheiros é $ 459.307,01.

## Q10: Qual o preço mínimo entre as casas com 3 quartos?

In [19]:
# verificando o preço mínimo entre as casas com 3 quartos
df[['price', 'bedrooms']].loc[df['bedrooms'] == 3].min()

price       89000.0
bedrooms        3.0
dtype: float64

**Resposta**: O preço mínimo entre as casas com 3 quartos é $ 89.000,00.

## Q11: Quantas casas possuem mais de 300 metros quadrados de sala de estar?

Para responder essa questão, será necessário fazer a conversão de pés para metros quadrados, bastando, para tanto, realizar a multiplicação por 0.0929.

In [20]:
# criando uma nova coluna para receber a conversão
df['m2_living'] = df['sqft_living'] * 0.0929

# verificando o resultado
df.head()

id             date     price  ...  sqft_living15  sqft_lot15  m2_living
0  7129300520  20141013T000000  221900.0  ...           1340        5650    109.622
1  6414100192  20141209T000000  538000.0  ...           1690        7639    238.753
2  5631500400  20150225T000000  180000.0  ...           2720        8062     71.533
3  2487200875  20141209T000000  604000.0  ...           1360        5000    182.084
4  1954400510  20150218T000000  510000.0  ...           1800        7503    156.072

[5 rows x 22 columns]

In [21]:
# verificando quantas casas possuem mais de 300 metros quadrados de sala de estar
df.loc[df['m2_living'] > 300].shape[0]

2251

**Resposta**: Existe 2.251 imóveis com mais de 300 metros quadrados de espaço interno.

## Q12: Quantas casas tem mais de 2 andares?

In [22]:
# verificando quantas casas têm mais de 2 andares
df.loc[df['floors'] > 2].shape[0]

780

**Resposta**: 780 imóveis possuem mais de 2 andares.

## Q13: Quantas casas tem vista para o mar?

Nessa base de dados as casas que não têm vista para o mar estão preenchidas com 0 e as que têm, com 1.

In [23]:
# verificando quantas casas têm vista para o mar
df['waterfront'].value_counts()

0    21273
1      163
Name: waterfront, dtype: int64

**Resposta**: 163 imóveis possuem vista para o mar.

## Q14: Das casas com vista para o mar, quantas tem 3 quartos?

In [24]:
# verificando quantas casas com vista para o mar tem 3 quartos
df.loc[(df['waterfront'] == 1) & (df['bedrooms'] == 3)].shape[0]

64

**Resposta**: Das casas com vista para o mar, 64 possuem 3 quartos.

## Q15: Das casas com mais de 300 metros quadrados de sala de estar, quantas tem mais de 2 banheiros?

In [25]:
# verificando quantas casas com mais de 300 metros têm mais de 2 banheiros
df.loc[(df['m2_living'] > 300) & (df['bathrooms'] > 2)].shape[0]

2194

**Resposta**: Das casas com mais de 300 metros quadrados, 2.194 possuem mais de 2 quartos.